### ISSUE TITLE
Dataloader and Dataset with extra node information

### ISSUE THREAD

**Issue by keyulu** [pyg-team/pytorch_geometric#2899]
Hi, I am implementing a dataloader and dataset to load from disk (many graphs), where each node not only has node feature but also extra information (e.g., weights that are used in loss but should not be treated as feature). 

From my understanding,  we first define the Dataset class by implementing the get(idx) function (for standard torch it's get_item), which loads from disk the graph data for idx, and makes a Data object from the loaded data. Then the torch_geometric Dataloader class will automatically deal with the sparse batching and indexing based on the Dataset. However, for each get() in Dataset, ideally I want to  return both Data and the node extra information. In this case, the default dataloader will have trouble with batching? Is there an easy way to deal with this? Thank you. 

Another related issue: is there an easy way to use torch_geometric with https://github.com/webdataset/webdataset? A package similar to TFRecord in Tensorflow that does sharding for faster loading from disk. Thanks!

**Comment by rusty1s** [comment]
For using `webdataset`, please see [here](https://github.com/rusty1s/pytorch_geometric/discussions/2864). It would be great to have an example of this.

Regarding your other questions: I suggest to put every relevant information into a single `Data` object. The `DataLoader` will then take care of mini-batching. Note that you can also return a `tuple` in `get(idx)`, however, for any non-`Data` object, the default PyTorch `collate_fn` will be used (which might fail in case when operating on graphs with varying sizes).

**Comment by FarzanT** [comment]
What if we have a multi-modal network, where one modality is processed via a GNN and others are ANNs/CNNs?
Can `DataLoader` handle both image/Tensor inputs and graph data? Should inputs for other modalities also be put in the `Data` object?
Thanks a lot!

**Comment by rusty1s** [comment]
You have three options here:
1. Just add the images to `data` as well with `__cat_dim__ = None`, see [here](). Otherwise, you will probably need to do a reshape.
2. Return both `data` and `img` values in `Dataset.get()`: This will use the `default_collate` for images:
```python
for data, img in loader:
    ...
```
3. Use two `DataLoader`s, one for PyG `data` objects and one for images. Note that you need to synchronize the random generator for this:
```python
for data, img in zip(graph_loader, img_loader):
    ...
```

**Comment by keyulu** [comment]
> You have three options here:
> 
> 1. Just add the images to `data` with `__cat_dim__ = None`, see [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html#batching-along-new-dimensions). Without modifying `__cat_dim__` behavior, you will probably need to do a reshape after mini-batching.
> 2. Return both `data` and `img` values in `Dataset.get()`: This will make use of the `default_collate` for images:
> 
> ```python
> for data, img in loader:
>     ...
> ```
> 
> 1. Use two `DataLoader`s, one for PyG `data` objects and one for images. Note that you will need to use the same random generator for this:
> 
> ```python
> for data, img in zip(graph_loader, img_loader):
>     ...
> ```

Is there a way to apply this to DataListLoader, which is needed for distributed training? Tried this but didn't work, thanks!!

**Comment by rusty1s** [comment]
The first approach should work with `DataListLoader` as well. Does it not work for you? What's the error message?

Alternatively, you can also make use of [`DistributedDataParallel` training](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/multi_gpu/distributed_batching.py), which is more straightforward to apply in this scenario.

**Comment by keyulu** [comment]
> The first approach should work with `DataListLoader` as well. Does it not work for you? What's the error message?
> 
> Alternatively, you can also make use of [`DistributedDataParallel` training](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/multi_gpu/distributed_batching.py), which is more straightforward to apply in this scenario.

Thank you, this is what I am looking for. Somehow previous posts say distributed data parallel is not supported, great to see it's now deployed!

**Comment by FarzanT** [comment]
I have tried setting `__cat_dim__` to `None`, but I have a list of multiple inputs for 2+ modalities. When I use torch `DataLoader` on my custom `Dataset` without graph data, I automatically get batches of shape `[Tensor(batch_size, modality_1), Tensor(batch_size, modality_2)]`. Now suppose one of these modalities becomes graph data. Using torch_geometric `DataLoader` I get `[batch_size * Tensor(modality_1, modality2)]`
In summary, I need each modality tensor to have the shape `(batch_size, modality_1_shape)`, then concatenated to a list.
Is there a specific `__cat_dim__` value to fix this, or should I overhaul my custom `Dataset`?

Thanks a lot for your help!

**Comment by rusty1s** [comment]
Can you show me a minimal example with dummy data to reproduce? Its' hard for me to see what's the exact input and desired output. Happy to help.

**Comment by FarzanT** [comment]
Here's a somewhat minimal example. In the first case, I create a Multi-Modal dataset without graph data, where a tuple of (list of tensors, tensor) is returned. The Pytorch DataLoader batches each Tensor within the list with the same tensor from other samples. In the second example where a custom `Data` class with `__cat_dim__ = None` is used inside a `data.Dataset` class, each list becomes a datapoint, which is not desired.

```python
import torch
import torch_geometric
from torch.utils import data
from torch_geometric.data import Data


class MultiModalDataSet(data.Dataset):
    def __init__(self):
        super(MultiModalDataSet, self).__init__()
        self.modality_1 = torch.rand((16, 32))
        self.modality_2 = torch.rand((16, 64))
        self.modality_3 = torch.rand((16, 48))

        self.targets = torch.rand((16, 1))

    def __len__(self):
        return self.modality_1.shape[0]

    def __getitem__(self, idx):
        return [self.modality_1[idx, ], self.modality_2[idx, ], self.modality_3[idx,]], self.targets


# Pytorch DataLoader batching/collation
my_data = MultiModalDataSet()
my_loader = data.DataLoader(my_data, batch_size=4)
my_iter = iter(my_loader)
torch_dataloader_sample = next(my_iter)

len(torch_dataloader_sample)  # 2, for data + targets
len(torch_dataloader_sample[0])  # 3, for 3 modalities
torch_dataloader_sample[0][0].shape  # modality 1 batched together
torch_dataloader_sample[0][1].shape  # modality 2 batched together
torch_dataloader_sample[0][2].shape  # modality 3 batched together


class MyGNNData(Data):
    def __cat_dim__(self, key, item):
        if key in ['modality_1', 'modality_2', 'modality_3', 'target']:
            return None
        else:
            return super().__cat_dim__(key, item)


class MultiModalGraphDataSet(data.Dataset):
    def __init__(self):
        super(MultiModalGraphDataSet, self).__init__()
        self.data = []
        for i in range(16):
            modality_1 = torch.rand((1, 32))
            modality_2 = torch.rand((1, 64))
            modality_3 = torch.rand((1, 48))
            target = torch.rand((1, 1))
            edge_index = torch.tensor([
               [0, 1, 1, 2],
               [1, 0, 2, 1],
            ])

            data = MyGNNData(edge_index=edge_index,
                             modalities=[modality_1, modality_2, modality_3],
                             target=target)
            self.data.append(data)

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return self.data[idx]


# Torch Geometric DataLoader batching/collation
my_data = MultiModalGraphDataSet()
my_loader = torch_geometric.data.DataLoader(my_data, batch_size=4)
my_iter = iter(my_loader)
torchgeo_dataloader_sample = next(my_iter)

len(torchgeo_dataloader_sample.modalities)  # 4, 1 for each sample/datapoint
len(torchgeo_dataloader_sample.modalities[0])  # 3, for 3 modalities
torchgeo_dataloader_sample.modalities[0][0].shape  # modality 1,2,3 batched together as 1 sample

```

**Comment by rusty1s** [comment]
I see. Currently, the `torch_geometric.data.DataLoader` cannot stack tensors appropriately when used inside a list. As an alternative, the following should yield the desired behaviour:
```python
data = MyGNNData(edge_index=edge_index, modality_1=modality_1,
                 modality_2=modality_2, modality_3=modality_3,
                 target=target)

```

**Comment by FarzanT** [comment]
Thank you, but since my Dataset outputs a different number of modalities depending on the scenario, I cannot directly assign attributes. I think I can still use the second option you provided here:

> 2. Return both `data` and `img` values in `Dataset.get()`: This will make use of the `default_collate` for images:
> 
> ```python
> for data, img in loader:
>     ...
> ```

And then pass this Dataset to a torch_geometric Dataloader.
Thanks for your time!

**Comment by rusty1s** [comment]
Note that classic PyTorch data loading will still fail if it tries to batch lists of modalities of different size.

**Comment by FarzanT** [comment]
Sorry, what I mean is that it can return different number modalities in separate training runs. In the same run, it will always return the same number of modalities. E.g. when we wanna compare (audio + image + text) to (audio + image) and (image + text) etc. I don't wanna create a separate `Dataset` for each possible combination.
Your solution above works though!
Thanks

**Comment by kshitijgundale** [comment]
Hi @rusty1s , I have labelled dataset that contains list edges, node features and piece of text associated with graph. I plan use to PyG for graph and BERT for text feature. Will it make sense if create a list of Data objects and simply assign the text as an attribute to every Data object ?

`data.text = text`

**Comment by rusty1s** [comment]
Yes, that is the recommended way.
